hibernate-core/src/main/java/org/hibernate/type/descriptor/java/GenericArrayTypeDescriptor.java

```diff
@@ -0,0 +1,358 @@
+	}
+
+	@Override
+	public MutabilityPlan<T[]> getMutabilityPlan() {
+		return super.getMutabilityPlan();
+	}
+
+	@Override
+	public String toString(T[] value) {
+		if ( value == null ) {
+			return null;
+		}
+		StringBuilder sb = new StringBuilder();
+		sb.append( '{' );
+		String glue = "";
+		for ( T v : value ) {
+			sb.append( glue );
+			if ( v == null ) {
+				sb.append( "null" );
+				glue = ",";
+				continue;
+			}
+			sb.append( '"' );
+			String valstr = this.componentDescriptor.toString( v );
+			// using replaceAll is a shorter, but much slower way to do this
+			for (int i = 0, len = valstr.length(); i < len; i ++ ) {
+				char c = valstr.charAt( i );
+				// Surrogate pairs. This is how they're done.
+				if (c == '\\' || c == '"') {
+					sb.append( '\\' );
+				}
+				sb.append( c );
+			}
+			sb.append( '"' );
+			glue = ",";
+		}
+		sb.append( '}' );
+		String result = sb.toString();
+		return result;
+	}
+
+	@Override
+	public T[] fromString(String string) {
+		if ( string == null ) {
+			return null;
+		}
+		java.util.ArrayList<String> lst = new java.util.ArrayList<>();
+		string = string.trim();
+		StringBuilder sb = null;
+		char lastchar = string.charAt( string.length() - 1 );
+		char firstchar = string.charAt( 0 );
+		if ( firstchar != '{' || lastchar != '}' ) {
+			throw new IllegalArgumentException( "Cannot parse given string into array of strings. First and last character must be { and }" );
+		}
+		int len = string.length();
+		int applen;
+		boolean inquote = false;
+		for ( int i = 1; i < len; i ++ ) {
+			char c = string.charAt( i );
+			char quote;
+			if ( c == '"' ) {
+				if (inquote) {
+					lst.add( sb.toString() );
+				}
+				else {
+					sb = new StringBuilder();
+				}
+				inquote = !inquote;
+				continue;
+			}
+			else if ( !inquote ) {
+				if ( Character.isWhitespace( c ) ) {
+					continue;
+				}
+				else if ( c == ',' ) {
+					// treat no-value between commas to mean null
+					if ( sb == null ) {
+						lst.add( null );
+					}
+					else {
+						sb = null;
+					}
+					continue;
+				}
+				else {
+					// i + 4, because there has to be a comma or closing brace after null
+					if ( i + 4 < len
+							&& string.charAt( i ) == 'n'
+							&& string.charAt( i + 1 ) == 'u'
+							&& string.charAt( i + 2 ) == 'l'
+							&& string.charAt( i + 3 ) == 'l') {
+						lst.add( null );
+						i += 4;
+						continue;
+					}
+					if (i + 1 == len) {
+						break;
+					}
+					throw new IllegalArgumentException( "Cannot parse given string into array of strings."
+																+ " Outside of quote, but neither whitespace, comma, array end, nor null found." );
+				}
+			}
+			else if ( c == '\\' && i + 2 < len && (string.charAt( i + 1 ) == '\\' || string.charAt( i + 1 ) == '"')) {
+				c = string.charAt( ++i );
+			}
+			// If there is ever a null-pointer here, the if-else logic before is incomplete
+			sb.append( c );
+		}
+		String[] objects = lst.toArray( new String[lst.size()] );
```

Comment:
In modern JVM, `list.toArray( new String[0] )` is always significantly faster than `list.toArray( new String[list.size()] )`, let alone much better code readability and maintainability. We might well go even further by using the various empty constants in ArrayHelper to avoid the performance cost of `new String[0]`.